### ISSUE TITLE
In-place ops on leaf Variables create reference cycles

### ISSUE THREAD

**Issue by glample** [pytorch/pytorch#839]
This script runs out of memory:

```python
for _ in range(100000):
    init = Variable(torch.cuda.FloatTensor(2, 32, 512).zero_())
    init[0] = Variable(torch.cuda.FloatTensor(32, 512).zero_())
```

**Comment by szagoruyko** [comment]
looks like any inplace tensor modification causes oom:

```python
for _ in range(100000):
    init = Variable(torch.cuda.FloatTensor(2, 32, 512).zero_())
    init *= 2
```

fails too

**Comment by apaszke** [comment]
Ok, so it seems that it's not really a memory leak, but a reference cycle. You're both modifying the leaf Variable in-place so the graph ends up looking like `init <- MulConstant <- init`, and nothing will be freed until the gc kicks in. If you only modify the loop to look like that:
```
for _ in range(100000):
    init = Variable(torch.cuda.FloatTensor(2, 32, 512).zero_())
    init = init + 0
    init *= 2
```
the memory usage will remain constant.

**Comment by apaszke** [comment]
Fixed in #852.
